homeassistant/components/synology_dsm/__init__.py

```diff
@@ -144,30 +247,233 @@ async def async_setup(self):
             device_token=self._entry.data.get("device_token"),
         )
 
+        await self._async_should_fetch_api()
+
         await self._hass.async_add_executor_job(self._fetch_device_configuration)
-        await self.update()
+        await self.async_update()
 
         self._unsub_dispatcher = async_track_time_interval(
             self._hass,
-            self.update,
+            self.async_update,
             timedelta(
                 minutes=self._entry.options.get(
                     CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
                 )
             ),
         )
 
+    async def _async_should_fetch_api(self):
+        """Determine if we should fetch each API, if one entity needs it."""
+        # Entities not added yet, fetch all
+        if not self.fetching_entities:
+            return
+
+        # Determine which if at least one entity uses specific API
+        self._with_security = False
+        self._with_storage = False
+        self._with_utilisation = False
+
+        for entity_unique_id in self.fetching_entities:
+            # Check if we should fetch specific APIs
+            if SynoCoreSecurity.API_KEY in entity_unique_id:
+                self._with_security = True
+                continue
+
+            if SynoStorage.API_KEY in entity_unique_id:
+                self._with_storage = True
+                continue
+
+            if SynoCoreUtilization.API_KEY in entity_unique_id:
+                self._with_utilisation = True
+                continue
+
```

Comment:
Much better, but its still a little brittle here as is now dependent on the unique id to never change which isn't currently covered by tests.

How about `self._api.fetching_entities.add(TYPE,UNIQUE_ID)` instead?

Then you can have a structure like this:
```
self.fetching_entities = {
    'security': { id1, id2, id3 },
    'storage': { id1, id2, id3 },
   ...
}
```

Then instead of iterating you can check
`if 'storage' in self.fetching_entities and len(self.fetching_entities['storage']):`

Reply:
It always depended on unique_id :sweat_smile: (see commit "_Do not fetch the API if all entries of it are disabled_")

But OK, will look at the dict way so it avoids the loop :blush:

Reply:
I think I got your idea, cool !